README.md
# TheAssistent

![logo](https://raw.githubusercontent.com/JeroenBL/TheAssistent/main/logo.png)

- [TheAssistent](#theassistent)
  - [Introduction](#introduction)
  - [GPT](#gpt)
  - [Requirements](#requirements)
    - [Create an OpenAI account and API key](#create-an-openai-account-and-api-key)
  - [Costs](#costs)
  - [How to use](#how-to-use)
    - [Installation](#installation)
    - [Version check](#version-check)
    - [Command palette](#command-palette)
  - [Settings](#settings)
  - [Commands](#commands)
  - [Examples](#examples)
    - [Explain code](#explain-code)
    - [Get a completion from the OpenAI API.](#get-a-completion-from-the-openai-api)
    - [Get a code review for the selected code](#get-a-code-review-for-the-selected-code)
  - [License](#license)
  - [Contributing](#contributing)

## Introduction

The `TheAssistent` extension is designed to interact with the OpenAI API through VSCode. While there are several PowerShell modules available that interact with the API, my personal workflow is centered around VSCode rather than a PowerShell console. For this reason, I developed the TheAssistent extension with most of its commands focused on PowerShell and creating Markdown documentation. 

> The name `TheAssistent` is temporary and subject to change in the future, as it is currently used as a development name.

> If you have suggestions or ideas, see: [Contributing](#contributing)

## GPT
GPT or Generative Pre-trained Transformer is a type of artificial intelligence developed by OpenAI. It is capable of generating human-like text by predicting the next word in a sequence based on the input it has been trained on. GPT-3, in particular, is one of the largest and most advanced language models to date, with the ability to perform a wide range of tasks, such as translation, summarization, question-answering, and more. 

## Requirements

- [ ] Account on: https://openai.com/api/
- [ ] Personal API key

### Create an OpenAI account and API key

1. Create an account on: https://openai.com/api/
2. Go to: https://platform.openai.com/
3. Click on: `Personal -> View API keys`
4. Click on: `Create new secret key`
5. Copy and paste your API key in the VSCode setting: `theassistent.APIKey`.

## Costs

OpenAI API's require payment. Upon creating an API key, new users receive a 18 dollar trial credit, which can be used to make API calls. The cost of API usage is calculated based on the number of tokens generated, with 1000 tokens equating to one cost unit. The number of tokens generated depends on the data sent to the API and the data received from it, and varies based on the model being used. More advanced models tend to be more expensive. You can use OpenAI's tokenizer to get an insight. Please refer to: https://platform.openai.com/tokenizer

## How to use

### Installation

To install this extension: 

1. Click on the extensions icon or press: `ctrl+shift+x` (`cmd+shift+x` on mac).
2. Click on the three dots `...` and select: `Install from VSIX`.
3. Browse to the folder where the VSIX file is downloaded.

### Version check

The first time the extension is used, a check will be performed to determine if a new version is available. If a new version is found, an informational message will be displayed indicating the version number and a download link.

> Note that this check will be performed each time you use the extension in a new VSCode instance.

### Command palette

All commands are available from the VSCode command palette by clicking on: `View -> Command palette` or press: `ctrl+shift+p` (`cmd+shift+p` on mac). 

To see all available commands, browse to: `The Assistent`.

### Context menu

Some commands are also available from the `context menu`. See the [Command Setion](#commands) for an overview.

> :exclamation: __The commands in the context menu are only available when text is selected.__

## Settings

The extension contributes to the following settings:

| Setting | Description |
| - | - |
| TheAssistent.APIKey | The ApiKey to authenticate to the OpenAI API |
| TheAssistent.MaxTokens | The maximum number of tokens to generate. The default value is set to: `500` |
| TheAssistent.Model | Name of the model to use. The default model is set to: `text-davinci-003` which is GPT's most advanced model |
| TheAssistent.Temperature | Controls the randomness of the generated text. `0` means more strict, `1` more creative. The default value is set to: `0` |
| TheAssistent.TopP | Controls how many different words or phrases the language model considers when it's trying to predict the next word. The default value is set to: `0` |

To modify the settings:

1. Click on: `File -> Preferences -> Settings` or press: `ctrl+,` (`cmd+,` on mac) and type: `The Assistent`.

For an overview of all available models, please refer to: https://platform.openai.com/docs/models/overview

## Commands 

The following commands are available:

| Command | Description | Settings | Where to find |
| - | - | - | - |
| theassistent.convertToAcademicEnglish | Converts the current line(s) to academic English | - Model: `text-davinci-003`<br> - Temperature: `1`<br> - MaxTokens: `700`<br> - TopP: `0.1` | - Command palette<br> - Context menu
| theassistent.convertToSpecificLanguage | Converts the current line(s) to the specified language | - Model: `text-davinci-003`<br> - Temperature: `0.3`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br>
| theassistent.getCompletionBasedOnUserPrompt | Get a completion from the OpenAI API | As specified: in `theassistent.configuration` | - Command palette<br>
| theassistent.convertToPowerShell | Converts code to PowerShell | - Model: `text-davinci-003`<br> - Temperature: `0.3`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br> - Context menu
| theassistent.explainCode | Explains the selected code | - Model: `text-davinci-003`<br> - Temperature: `1`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br> - Context menu
| theassistent.createPowerShellRegex | Creates a new regex based on the provided details | - Model: `text-davinci-003`<br> - Temperature: `1`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br>
| theassistent.addCodeComments | Add code comments to the selected PowerShell code | - Model: `text-davinci-003`<br> - Temperature: `1`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br>
| theassistent.requestCodeReview | Request a code review for the selected PowerShell code | - Model: `text-davinci-003`<br> - Temperature: `0`<br> - MaxTokens: `undefined`<br> - TopP: `1` | - Command palette<br> - Context menu
| theassistent.getMeetingSummary | Converts your short descriptive notes to a first hand account | As specified in: `theassistent.configuration` | - Command palette<br>
| theassistent.explainError | Explains the error and provides solutions and ways to debug | As specified in: `theassistent.configuration` | - Command palette<br> - Context menu

> :exclamation: If a setting is set to `undefined` the default value will be used.

> For more information on the settings, please refer to: [Settings](#settings)

## Examples

### Explain code

TheAssistent can help you understand code. 

1. Select the code that you would like to have explained.
2. Open the command palette by clicking on: `View -> Command palette` or press: `ctrl+shift+p` (`cmd+shift+p` on mac). 
3. Browse to: `The Assistent`.
4. Look for: `Explains the selected code` and press: `enter`.
5. A new window will open with the results.

![explainCode](https://raw.githubusercontent.com/JeroenBL/TheAssistent/main/assets/explainCodeUpdated.gif)

### Get a completion from the OpenAI API.

Sometimes, you just want to ask a question. Whether that's about code or something random like: `Create person data for 2 persons. I need the following fields: userId, employeeId, firstName, lastName, eMailAddress, Title, phoneNumber. Return the data as JSON.`

1. Open the command palette by clicking on: `View -> Command palette` or press: `ctrl+shift+p` (`cmd+shift+p` on mac). 
2. Browse to: `The Assistent`.
3. Look for: `Get a completion from the OpenAI API` and press: `enter`.
4. Specify the prompt and press: `enter`.

![prompt](https://raw.githubusercontent.com/JeroenBL/TheAssistent/main/assets/getCompletionPersons.gif)

> :bulb: The request to the OpenAI API is made using the settings in the `The Assistent` configuration section.

### Get a code review for the selected code

Code reviews are always helpful to make sure your code is correct and does not contain bugs. But, in some cases, a colleague might not be available. `The Assistent` can assist you and provide an in-depth review.

1. Select the code that you would like to have reviewed.
2. Open the command palette by clicking on: `View -> Command palette` or press: `ctrl+shift+p` (`cmd+shift+p` on mac). 
3. Browse to: `The Assistent`.
4. Look for: `Request a code review for the selected PowerShell code` and press: `enter`.

The review is always split into four sections:

- Pros
- Cons
- Bugs
- Improvements

![prompt](https://raw.githubusercontent.com/JeroenBL/TheAssistent/main/assets/codeReview.gif)

## License

This extension is released under the [MIT License](https://github.com/JeroenBL/TheAssistent/blob/main/LICENSE).

## Contributing

Find a bug or have an idea! Open an issue or submit a pull request!

**Enjoy!**

CHANGELOG.md
# Change Log

## 1.0.3 - 08-03-2023

### Fixed

- Fixed issue that all regex where wrapped with backticks.

### Changed

- Changed the prompt that was send to be rendered as markdown in the htmlview.

### Added

- Added command: `explainError` to will explain errors that occured in your code.
- Added commands: `convertToAcademicEnglish`, `convertToPowerShell`, `explainCode`, `requestCodeReview` and `explainError` to the context menu but only to
show when text is selected in the active editor.

## 1.0.2 - 03-03-2023

### Fixed

- Fixed issue that the completion text was not wrapped.
- Fixed issue that original prompt was not aligned.

## 1.0.1 - 02-03-2023

### Fixed

- Fixed issue that, when using a regex, the prompt send to GPT was not vissible in the webview panel.
- Fixed bug that the `selectedText` was given to the `getCompletionResult` function, but never used.
- Removed timeout that extension did not load until VSCode was fully loaded.

### Changes

- Added a `Copy` button to the webview panel.
- Render the original prompt as a `blockquote`.
- Added the option to cancel the API request to OpenAI with a `cancel` button on the progress bar.

## 1.0.0 - 26-02-2023

1.0.0 release of: `The Assistent`.

### Fixed

### Changes

### Added commands

## 0.0.8 - 26-02-2023

### Fixed

- Using the function without a completion, but click somewhere in the editor, results in an error.
- Extension fails when VSCode is not fully loaded

### Changes

- Updated documentation.
- Fixed typo's.

### Added commands

## 0.0.7 - 25-02-2023

### Fixed

### Changes

- Updated name of the configuration settings from: `theassistent configuration` to: `TheAssistent`.
- Refactored code and removed duplicates.
- Updated readme.

### Added commands

- `theassistent.requestCodeReview`
- `theassistent.getMeetingSummary`

## 0.0.6 - 24-02-2023

### Fixed

### Changes

- Updated check to see if a new version is available. If so, a `window.showInformationMessage` will be displayed.
- Removed package `request-promise-native`. 
- Refactored `getLatestReleaseInfo` function to use axios instead of `request-promise-native`.
- Added progress bar to the `getAICompletion` function to, visually show that an API call is being made.
- Changed `theassistent.askGPT` to always use the settings from the `theassistent` configuration.
- Changed name of: `theassistent.askGPT` to: `theassistent.getCompletionBasedOnUserPrompt`.
- Removed `theassistent.anonymizeCode`.
- Updated readme.

### Added commands

## 0.0.5 - 23-02-2023

### Fixed

### Changes

- Renamed all commands in documentation to the actual name used in the extension.

### Added commands

- theassistent.anonymizeCode

## 0.0.4 - 23-02-2023

### Fixed

- Fixed configuration settings that `topP` was written in PascalCase.
- Fixed `@` character in HTML view for the `Original text`.

### Changes

- Added `markdown-it` to render all code blocks in PowerShell Markdown in the HTML view.
- Updated prompts for commands: `theassistent.convertToPowerShell` and `theassistent.createRegex`
- Changed command name: `theassistent.createRegex` to `theassistent.createPowerShellRegex`
- Added categories: `Productivity` and `Text editing` in the `contributes` section.

### Added commands

- theassistent.addCodeComments

## 0.0.3 - 23-02-2023

### Fixed

### Changes

- Updated title to `TheAssistent` for easy seaching in the command window.

### Added commands

## 0.0.2 - 23-02-2023

### Fixed

- Solved bug that no all selected lines got translated correctly.

### Changes

The following command has been renamed:

- ConvertTo-SpecificLanguage (previously: ConvertToAnyLangauge)

### Added commands

- theassistent.convertToPowerShell
- theassistent.explainCode 
- theassistent.createPowerShellRegex 

## 0.0.1 - 22-02-2023

Initial release of TheAssistent

### Added commands

- theassistent.convertToAcademicEnglish
- theassistent.convertToSpecificLanguage
- theassistent.askGPT

src/extension.ts
import * as vscode from 'vscode';
import * as markdownit from 'markdown-it';

const axios = require("axios");
const packageJson = require('../package.json');
const md = new markdownit();

// Gets called when the extension is activated
export async function activate(context: vscode.ExtensionContext) {
	const latestVersion = await getLatestReleaseInfo('JeroenBL', 'TheAssistent');
	if (latestVersion && latestVersion.version !== packageJson.version) {
		await vscode.window.showInformationMessage(`Version: [${latestVersion.version}] of TheAssistent is now available on: [Github](https://github.com/JeroenBL/TheAssistent/releases/latest)`);
	}

	registerCommands();
}

export function deactivate() { }

// Registers all available commands from the extension in VSCode
function registerCommands() {
	vscode.commands.registerCommand('theassistent.convertToAcademicEnglish', registerCommandConvertToAcademicEnglish);
	vscode.commands.registerCommand('theassistent.convertToSpecificLanguage', registerCommandConvertToSpecificLanguage);
	vscode.commands.registerCommand('theassistent.convertToPowerShell', registerCommandConvertToPowerShell);
	vscode.commands.registerCommand('theassistent.explainCode', registerCommandExplainCode);
	vscode.commands.registerCommand('theassistent.addCodeComments', registerCommandAddCodeComments);
	vscode.commands.registerCommand('theassistent.createPowerShellRegex', registerCommandCreatePowerShellRegex);
	vscode.commands.registerCommand('theassistent.getCompletionBasedOnUserPrompt', registerCommandGetCompletionBasedOnUserPrompt);
	vscode.commands.registerCommand('theassistent.requestCodeReview', registerCommandRequestCodeReview);
	vscode.commands.registerCommand('theassistent.getMeetingSummary', registerCommandGetMeetingSummary);
	vscode.commands.registerCommand('theassistent.explainError', registerCommandExplainError);
}

// Registers the command for: [ConvertToAcademicEnglish]
async function registerCommandConvertToAcademicEnglish() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);
  
	const prompt = `I want you to act as a language translator. You will translate my text to academic English and correct grammar and spelling mistakes. My text:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 1;
	const maxTokens = 700;
	const topP = 0.1;
	
	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [ConvertToSpecificLanguage]
async function registerCommandConvertToSpecificLanguage() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);
	const inputLanguage = await vscode.window.showInputBox({
		prompt: 'Specify language',
		value: 'Spanish'
	});

	const prompt = `I want you to act as an ${inputLanguage} translator, correct spelling and improve my text. You will translate my text and answer in the corrected and improved version of my text in ${inputLanguage}. I want you to replace my simplified A0-level words and sentences with more beautiful and elegant, upper level ${inputLanguage} words and sentences. Keep the meaning same, but make them more literary. I want you to only reply with the correction, improvements and nothing else. My text:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 0.3;
	const maxTokens = undefined;
	const topP = 1;

	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [ConvertToPowerShell]
async function registerCommandConvertToPowerShell() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `I want you to convert the following code to PowerShell:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 0.3;
	const maxTokens = undefined;
	const topP = 1;

	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [ExplainCode]
async function registerCommandExplainCode() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `I want you to explain the following code:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 1;
	const maxTokens = undefined;
	const topP = 1;

	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [AddCodeComments]
async function registerCommandAddCodeComments() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `Add code comments for each line for the following PowerShell code:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 1;
	const maxTokens = undefined;
	const topP = 1;
	
	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [createPowerShellRegex]
async function registerCommandCreatePowerShellRegex() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);
	const input = await vscode.window.showInputBox({
		prompt: 'Specify the PowerShell regex you need',
		value: 'a regex that matches a url'
	});
	const prompt = `Write a regex using PowerShell match for:\n\n${input}\n\n and only return a concise PowerShell code example that clearly shows how to use it. Do not wrap the result in backticks.`;
	const model = "text-davinci-003";
	const temperature = 1;
	const maxTokens = undefined;
	const topP = 1;
	
	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [getCompletionBasedOnUserPrompt]
async function registerCommandGetCompletionBasedOnUserPrompt() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const prompt = await vscode.window.showInputBox({ prompt: 'Specify prompt' });
	
	if (prompt !== undefined && prompt !== '') {
		await getCompletionResult(prompt, undefined, undefined, undefined, undefined);
	}
}

// Registers the command for: [requestCodeReview]
async function registerCommandRequestCodeReview() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `I want you to act as a code reviewer. I will provide you with some PowerShell code and you will do an in-depth review - including pros, cons, bugs and improvements. My code:\n\n${selectedText}`;
	const model = "text-davinci-003";
	const temperature = 0;
	const maxTokens = undefined;
	const topP = 1;
	
	await getCompletionResult(prompt, model, temperature, maxTokens, topP);
}

// Registers the command for: [getMeetingSummary]
async function registerCommandGetMeetingSummary() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `Convert my short hand notes into a first-hand account of the meeting. My notes:\n\n${selectedText}`;
	
	await getCompletionResult(prompt, undefined, undefined, undefined, undefined);
}

// Registers the command for: [explainError]
async function registerCommandExplainError() {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		return;
	}

	const selection = editor.selection;
	const selectedText = editor.document.getText(selection);

	const prompt = `I want you to provide a clear explanation of the following error and also provide five solutions and ways to debug the error:\n\n${selectedText}`;
	
	await getCompletionResult(prompt, undefined, undefined, undefined, undefined);
}

// Async function that invokes the 'postAICompletion' function and creates the HTML view
async function getCompletionResult(prompt?: string, model?: string, temperature?: number, maxTokens?: number, topP?: number) {
	try {
		const response = await postAICompletion(prompt, model, temperature, maxTokens, topP);
		await createHTMLView(response, prompt);
	} catch (error) {
		console.error(error);
		throw error;
	}
}

// Async function that actually posts the GPT completion to the OpenAI API
async function postAICompletion(prompt?: string, model?: string, temperature?: number, maxTokens?: number, topP?: number): Promise<string> {
	// Get current configuration settings
	const configuration = vscode.workspace.getConfiguration("theassistent") as vscode.WorkspaceConfiguration;

	if (!model) { model = configuration.model; }
	if (!temperature) { temperature = configuration.temperature; }
	if (!maxTokens) { maxTokens = configuration.maxTokens; }
	if (!topP) { topP = configuration.topP; }

	const requestBody = {
		prompt: prompt,
		model: model,
		temperature: temperature,
		max_tokens: maxTokens,
		top_p: topP
	};

	const headers = {
		Authorization: `Bearer ${configuration.apiKey}`
	};

	// Show progress bar while making API call
	return vscode.window.withProgress({
		location: vscode.ProgressLocation.Notification,
		title: 'TheAssistent',
		cancellable: true // set cancellable option to true
	}, async (progress, token) => { // add token parameter to function
		token.onCancellationRequested(() => {
			// handle cancellation request
			vscode.window.showInformationMessage('API call cancelled');
		});
		progress.report({ message: 'Getting completion result...' });
		const response = await axios.post('https://api.openai.com/v1/completions', requestBody, { headers });
		progress.report({ message: 'Completion result ready...' });
		return response;
	});
}
  
// Async function that creates an HTML view containig the result
async function createHTMLView(result: any, lineText?: any) {
	const plainTextResult = result.data.choices[0].text;
	const htmlResult = `
	<div style="font-family: sans-serif; font-size: 14px;">
   	   ${md.render(lineText)}
	  <br />
	  <h2>Results:</h2>
	  <pre style="white-space: pre-wrap;">${plainTextResult}</pre>
	  <br />
	  <button id="copy-button" style="background-color: #4CAF50; color: white; border: none; padding: 10px 20px; text-align: center; text-decoration: none; display: inline-block; font-size: 16px; margin: 4px 2px; cursor: pointer;">Copy</button>
	</div>
	<body>
	  <script>
		  const vscode = acquireVsCodeApi();
		  const copyButton = document.querySelector("#copy-button");
	
		  // Add click event listener to copy button
		  copyButton.addEventListener("click", () => {
		  vscode.postMessage({ command: "copy" });
		  });
		  
		  vscode.postMessage({ command: "init" });
	  </script>
	</body>`;

	const panel = vscode.window.createWebviewPanel(
	  "Completion",
	  "Completion result",
	  vscode.ViewColumn.Beside,
	  {
		enableScripts: true // Enable JavaScript in the webview
	  }
	);
  
	panel.webview.html = htmlResult;

	// Add click event listener to copy button
	panel.webview.onDidReceiveMessage((message) => {
		if (message.command === "copy") {
		  let textToCopy = plainTextResult;
		  if (textToCopy.startsWith("<pre><code>") && textToCopy.endsWith("</code></pre>")) {
			textToCopy = textToCopy.slice(11, -13);
		  }
		  vscode.env.clipboard.writeText(textToCopy).then(() => {
			vscode.window.showInformationMessage("Text copied to clipboard!");
		  });
		}
	});

	// Add script to webview to send message when button is clicked
	panel.webview.postMessage({ command: "init" });
}

// Checks if the installed extension is the latest version and displays a message if its not
async function getLatestReleaseInfo(owner: string, repo: string) {
	const url = `https://api.github.com/repos/${owner}/${repo}/releases/latest`;
	const headers = {
		'User-Agent': 'axios'
	};

	try {
		const response = await axios.get(url, { headers });
		return {
			version: response.data.tag_name,
			url: response.data.html_url
		};
	} catch (error) {
		console.error(error);
		return null;
	}
}